### ISSUE TITLE
🐛 Biome encountered an unexpected error - Biome encountered an unexpected error

### ISSUE THREAD

**Issue by dougg0k** [biomejs/biome#6382]
### Environment information

```block
CLI:
  Version:                      2.0.0
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                8
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.16.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.12.1

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Workspace:
  Open Documents:               0
```

### What happened?

Not sure what caused it.

```
2025-06-18 09:16:14.687 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.


2025-06-18 09:16:21.644 [info] [Error - 9:16:21 AM] Server process exited with code 0.
2025-06-18 09:16:21.645 [info] [Info  - 9:16:21 AM] Connection to server got closed. Server will restart.
2025-06-18 09:16:21.645 [info] true
2025-06-18 09:16:21.707 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.


2025-06-18 09:16:21.709 [info] [Error - 9:16:21 AM] Server process exited with code 0.
2025-06-18 09:16:21.709 [info] [Info  - 9:16:21 AM] Connection to server got closed. Server will restart.
2025-06-18 09:16:21.709 [info] true
2025-06-18 09:16:21.771 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.


2025-06-18 09:16:21.772 [info] [Error - 9:16:21 AM] Server process exited with code 0.
2025-06-18 09:16:21.773 [info] [Info  - 9:16:21 AM] Connection to server got closed. Server will restart.
2025-06-18 09:16:21.773 [info] true
2025-06-18 09:16:21.834 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.


2025-06-18 09:16:23.327 [info] [Error - 9:16:23 AM] Server process exited with code 0.
2025-06-18 09:16:23.327 [info] [Info  - 9:16:23 AM] Connection to server got closed. Server will restart.
2025-06-18 09:16:23.327 [info] true
2025-06-18 09:16:23.391 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.


2025-06-18 09:16:23.392 [info] [Error - 9:16:23 AM] Server process exited with code 0.
2025-06-18 09:16:23.393 [info] [Error - 9:16:23 AM] The biome server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

### Expected result

To not crash the server.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by dougg0k** [comment]
Is there any way to disable the VSCode popups / toatsts, the ones from the bottom right, ideally from errors happening in biome?

I only saw the option to disable info and warnings, not errors.

**Comment by cjkihl** [comment]
I ran into the same error. Switching from Typescript Vscode version 5.7.2 to workspace version 5.8.3 fixed the error

<img width="587" alt="Image" src="https://github.com/user-attachments/assets/ae526d62-d7d5-4ce9-a62b-3ec4c9867786" />

**Comment by dougg0k** [comment]
Here both versions are on `5.8.3`.

**Comment by ematipico** [comment]
@dougg0k you really have to provide a reproduction, even the Biome message tells you to do so. We can't debug and help you without a consistent way to reproduce the issue.

**Comment by dougg0k** [comment]
If I find out the cause, I will, otherwise, if anyone else who gave the thumbs up, freel free to share a reproduction, if you find the actual reason.

---

This happened multiple times yesterday, but I was not sure of what caused it.

**Comment by dougg0k** [comment]
@ematipico 

Here is one way I found how to reproduce. And it only happens in a project that has `biome.json`.

VScode was opened two times, to show that the cause was not keepassxc secret service prompt. But actually related to vscode `settings.json`. Unsure of why, as it is properly formatted and it was not happening before `2.0`.

https://github.com/user-attachments/assets/3a11029d-9b77-4a7c-b4e5-7541abf8ea15

---

I said that about `biome.json`, because I did the same thing in a project without it. With `settings.json` pre-open or not, and it didnt happen, even though the `extension` were enabled.

```
{
	"$schema": "https://biomejs.dev/schemas/2.0.0/schema.json",
	"assist": {
		"actions": {
			"source": {
				"organizeImports": "on",
				"useSortedAttributes": "on",
				"useSortedKeys": "on",
				"useSortedProperties": "on"
			}
		}
	},
	"files": {
		"includes": ["**/src/**/*.ts", "**/src/**/*.tsx", "!**/node_modules"]
	},
	"formatter": {
		"enabled": true
	},
	"linter": {
		"enabled": true,
		"rules": {
			"correctness": {
				"noUnusedImports": "off",
				"useExhaustiveDependencies": "off"
			},
			"style": {
				"noInferrableTypes": "error",
				"noParameterAssign": "error",
				"noUnusedTemplateLiteral": "error",
				"noUselessElse": "error",
				"useAsConstAssertion": "error",
				"useDefaultParameterLast": "error",
				"useEnumInitializers": "error",
				"useNumberNamespace": "error",
				"useSelfClosingElements": "error",
				"useSingleVarDeclarator": "error"
			},
			"suspicious": {
				"noConsole": "error",
				"noExplicitAny": "off"
			}
		}
	}
}

```

**Comment by eikowagenknecht** [comment]
For me, it starts with some other error first and then progresses to this one:

```
2025-06-20 08:34:05.652 [info] [Info  - 8:34:05 AM] Server initialized with PID: 23644
2025-06-20 08:34:12.305 [info] [Info  - 8:34:12 AM] Server initialized with PID: 9324
2025-06-20 08:34:29.309 [info] [Info  - 8:34:29 AM] Server initialized with PID: 9324
2025-06-20 08:34:31.341 [info] [Error - 8:34:31 AM] Server process exited with code 1.
2025-06-20 08:34:38.528 [info] [Error - 8:34:38 AM] Client biome: connection to server is erroring.
write EPIPE
Shutting down server.
2025-06-20 08:34:38.528 [info] [Error - 8:34:38 AM] Request textDocument/codeAction failed.
2025-06-20 08:34:38.528 [info]   Message: write EPIPE
  Code: -32099 
2025-06-20 08:34:38.529 [info] [Error - 8:34:38 AM] Client biome: connection to server is erroring.
Cannot call write after a stream was destroyed
2025-06-20 08:34:38.529 [info] [Error - 8:34:38 AM] Stopping server failed
2025-06-20 08:34:38.529 [info]   Message: Cannot call write after a stream was destroyed
  Code: -32099 
2025-06-20 08:34:38.529 [info] [Error - 8:34:38 AM] Stopping server failed
2025-06-20 08:34:38.529 [info]   Message: Cannot call write after a stream was destroyed
  Code: -32099 
2025-06-20 08:34:38.537 [info] [Error - 8:34:38 AM] Server process exited with code 0.
2025-06-20 08:34:55.118 [info] [Info  - 8:34:55 AM] Server initialized with PID: 25844
2025-06-20 08:35:34.626 [info] [Info  - 8:35:34 AM] Server initialized with PID: 25844
2025-06-20 08:35:36.632 [info] [Error - 8:35:36 AM] Server process exited with code 1.
2025-06-20 08:36:00.166 [info] [Info  - 8:36:00 AM] Connection to server got closed. Server will restart.
2025-06-20 08:36:00.166 [info] true
2025-06-20 08:36:00.173 [info] [Error - 8:36:00 AM] Server process exited with code 0.
2025-06-20 08:36:00.315 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates\biome_service\src\settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.
```

biome.json file, biome 1.9.4:
```json
{
  "$schema": "https://biomejs.dev/schemas/1.9.4/schema.json",
  "organizeImports": {
    "enabled": true
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true
    }
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "space",
    "lineEnding": "lf"
  },
  "vcs": {
    "enabled": true,
    "clientKind": "git",
    "useIgnoreFile": true,
    "defaultBranch": "main"
  },
  "files": {
    "ignore": ["tsconfig.app.json"]
  }
}
```

I haven't figured out a pattern how to trigger this reliably though, so I can't provide great reproductions either.

**Comment by ematipico** [comment]
Ah, I see that the settings.json file is **outside** of your project. Maybe that's the reason? I'll see what I can do to reproduce it later

**Comment by OneOfOne** [comment]
Same issue here, however the biome.json file is in the root of my project  with tsconfig, etc.

**Comment by junlarsen** [comment]
I have made a MRE for Helix as the LSP client (I'm not a vscode user, sorry), but I'm assuming the same would apply for VSCode https://github.com/junlarsen/biome-lsp-crash-repro

**Comment by dougg0k** [comment]
> Ah, I see that the settings.json file is **outside** of your project. Maybe that's the reason? I'll see what I can do to reproduce it later

Yes @ematipico

**Comment by OneOfOne** [comment]
Any workarounds for this bug besides downgrading biome?

**Comment by dougg0k** [comment]
I dont see this happening anymore, at least not in `2.1.2`. 

If someone else confirms, issue can be closed. @ematipico

**Comment by OneOfOne** [comment]
It's still happening with 2.1.2 for me in neovim:

```
[ERROR][2025-07-17 13:47:33] /usr/share/nvim/runtime/lua/vim/lsp/log.lua:151	"rpc"	"biome"	"stderr"	"Biome encountered an unexpected error\n\nThis is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:\n\nSource Location: crates/biome_service/src/settings.rs:78:14\nThread Name: tokio-runtime-worker\nMessage: You must have at least one workspace.\n\n"
```

```
❯ biome --version
Version: 2.1.2
```

biome.json:

```json
{
	"$schema": "node_modules/@biomejs/biome/configuration_schema.json",
	"formatter": {
		"enabled": true,
		"formatWithErrors": true,
		"indentStyle": "tab",
		"indentWidth": 4,
		"lineWidth": 140
	},
	"vcs": {
		"enabled": true,
		"clientKind": "git",
		"useIgnoreFile": true
	},
	"javascript": {
		"formatter": {
			"quoteStyle": "single",
			"jsxQuoteStyle": "single"
		},
		"globals": ["__DEV__"]
	},
	"files": {},
	"assist": {
		"enabled": true,
		"actions": {
			"source": {
				"organizeImports": {
					"level": "on",
					"options": {
						"groups": [
							":NODE:",
							":BUN:",
							"react",
							"*react-native*",
							"*expo*",
							":PACKAGE:",
							":BLANK_LINE:",
							{ "type": false, "source": ":ALIAS:" }
						]
					}
				}
			}
		}
	},
	"linter": {
		"enabled": true,
		"rules": {
			"complexity": {
				"noUselessFragments": "off",
				"noExcessiveCognitiveComplexity": "off"
			},
			"style": {
				"noParameterAssign": "off",
				"noNonNullAssertion": "off",
				"useNamingConvention": "off",
				"useFilenamingConvention": "off",
				"useBlockStatements": "off",
				"noDefaultExport": "off",
				"noImplicitBoolean": "off",
				"useExplicitLengthCheck": "off",
				"noNamespace": "off"
			},
			"suspicious": {
				"noExplicitAny": "off",
				"noArrayIndexKey": "off",
				"noEmptyBlockStatements": "off",
				"useAwait": "off",
				"noConsole": "off"
			},
			"correctness": {
				"useExhaustiveDependencies": "off",
				"useImportExtensions": "off",
				"noUndeclaredDependencies": "off",
				"noNodejsModules": "off"
			},
			"performance": {
				"noReExportAll": "off",
				"noAccumulatingSpread": "off",
				"noBarrelFile": "off",
				"useTopLevelRegex": "off"
			},
			"security": {
				"noDangerouslySetInnerHtml": "off"
			}
		}
	}
}
```

**Comment by OneOfOne** [comment]
seems like setting `"assist": {"enabled": false}` fixes the issue.

**Comment by dougg0k** [comment]
I have plenty of assist settings. Still no issue. 

Before, everytime I opened vscode settings, it triggered the error.

```
	"assist": {
		"actions": {
			"source": {
				"organizeImports": "on",
				"useSortedAttributes": "on",
				"useSortedKeys": "off",
				"useSortedProperties": "on"
			}
		}
	},
```

**Comment by OneOfOne** [comment]
It seems like a bug in mason using the wrong version.

**Comment by ematipico** [comment]
So, I can reproduce the issue with Helix, but I think the issue is that helix doesn't run the version installed inside `package.json`, but it runs the global installed version. My global version is outdated, and it throws the error.

**Comment by ematipico** [comment]
Weirder and weirder: helix keeps trying to use biome v1.9.4 but I have no idea where it's pulling it from...

**Comment by Steviegt** [comment]
I've got the Same problem. 
The error occurs in nvim. When I delete the biome.jsonc everything is fine.

I'm running on biome 2.1.2 right now but :LspInfo shows biome version 1.9.4.

I'm using the default configuration from nvim-lspconfig on the newest version.

Does someone have a clue where the 1.9.4 version comes from?

**Comment by ematipico** [comment]
I can't reproduce it anymore

**Comment by junlarsen** [comment]
I managed to reproduce it today - it seems that Helix wants to use a globally available `biome` executable. Updating the global biome (perhaps installed through `pnpm i -g`) will fix it.

In any case the problem does not reproduce on latest
